tx-parser/src/lib.rs

```diff
@@ -106,18 +121,23 @@ def_matches_namespaced_keyword!(Tx, literal_db_retract, "db", "retract");
 
 def_parser!(Tx, add_or_retract, Entity, {
     vector().of_exactly(
+        // TODO: This commits as soon as it sees :db/{add,retract}, but could use an improved error message.
         (Tx::literal_db_add().map(|_| OpType::Add).or(Tx::literal_db_retract().map(|_| OpType::Retract)),
-             Tx::entid_or_lookup_ref_or_temp_id(),
-             Tx::entid(),
-             Tx::atom_or_lookup_ref_or_vector())
-        .map(|(op, e, a, v)| {
-            Entity::AddOrRetract {
-                op: op,
-                e: e,
-                a: a,
-                v: v,
-            }
-        }))
+          try((Tx::entid_or_lookup_ref_or_temp_id(),
+               Tx::forward_entid(),
+               Tx::atom_or_lookup_ref_or_vector()))
+          .or(try((Tx::atom_or_lookup_ref_or_vector(),
```

Comment:
I'm suspicious of the cost of `try`, especially when the two patterns here are so similar, and the disambiguation is often done by the second term — in the success case of a backward attribute you always parse the entity twice.

The code's already written, but I would probably have done this by matching the most general pattern and then filtering. Are my instincts wrong?